src/java.base/share/classes/java/io/ObjectInputFilter.java

```diff
@@ -48,50 +53,163 @@
- *     ObjectInputFilter serialFilter = ObjectInputFilter.Config.getSerialFilter();
- *     if (serialFilter != null) {
- *         ObjectInputFilter.Status status = serialFilter.checkInput(info);
- *         if (status != ObjectInputFilter.Status.UNDECIDED) {
- *             // The system-wide filter overrides this filter
- *             return status;
+ * public static final class FilterInThread implements BinaryOperator<ObjectInputFilter> {
+ *
+ *     private final ThreadLocal<ObjectInputFilter> filterThreadLocal = new InheritableThreadLocal<>();
+ *
+ *     // Construct a FilterInThread deserialization filter factory.
+ *     public FilterInThread() {}
+ *
+ *     // Returns a composite filter of the static JVM-wide filter, a thread-specific filter,
+ *     // and the stream-specific filter.
+ *     public ObjectInputFilter apply(ObjectInputFilter curr, ObjectInputFilter next) {
+ *         if (curr == null) {
+ *             // Called from the OIS constructor or perhaps OIS.setObjectInputFilter with no previous filter
+ *             // Prepend next to the threadFilter, both may be null or non-null
+ *             var filter = filterThreadLocal.get();
+ *             if (filter != null) {
+ *                 // Prepend a filter to assert that all classes have been Allowed or Rejected
+ *                 filter = filter.rejectUndecidedClass();
+ *             }
+ *             if (next != null) {
+ *                 // Prepend the next filter to the thread filter, if any
+ *                 // Initially this would be the static JVM-wide filter passed from the OIS constructor
+ *                 // Append the filter to reject all UNDECIDED results
+ *                 filter = next.merge(filter).rejectUndecidedClass();
+ *             }
+ *             return filter;
+ *         } else {
+ *             // Called from OIS.setObjectInputFilter with a current filter and a stream-specific filter.
+ *             // The curr filter already incorporates the thread filter and static JVM-wide filter
+ *             // and rejection of undecided classes
+ *             // Use the current filter or prepend the stream-specific filter and recheck for undecided
+ *             return (next == null) ? curr : next.merge(curr).rejectUndecidedClass();
  *         }
  *     }
- *     if (info.serialClass() != null &&
- *         Remote.class.isAssignableFrom(info.serialClass())) {
- *         return Status.REJECTED;      // Do not allow Remote objects
+ *
+ *     // Applies the filter to the thread and invokes the runnable.
+ *     public void doWithSerialFilter(ObjectInputFilter filter, Runnable runnable) {
+ *         var prevFilter = filterThreadLocal.get();
+ *         try {
+ *             filterThreadLocal.set(filter);
+ *             runnable.run();
+ *         } finally {
+ *             filterThreadLocal.set(prevFilter);
+ *         }
  *     }
- *     return Status.UNDECIDED;
  * }
- *}</pre>
+ * }</pre>
+ * <h2>Using the Filter Factory</h2>
```

Comment:
Should this be rather a subsection of the example above? Otherwise it gives the impression that the `doWithSerialFilter` method is provided or invoked by the JDK.